x/delayedack/hooks.go

```diff
@@ -32,24 +33,70 @@ func (im IBCMiddleware) FinalizeRollappPackets(ctx sdk.Context, rollappID string
 	// Get the packets for the rollapp until height
 	logger.Debug("Finalizing IBC rollapp packets", "rollappID", rollappID, "state end height", stateEndHeight, "num packets", len(rollappPendingPackets))
 	for _, rollappPacket := range rollappPendingPackets {
-		logger.Debug("Finalizing IBC rollapp packet", "rollappID", rollappID, "sequence", rollappPacket.Packet.GetSequence(), "destination channel", rollappPacket.Packet.GetDestChannel())
+		logger.Debug("Finalizing IBC rollapp packet", "rollappID", rollappID, "sequence", rollappPacket.Packet.GetSequence(), "destination channel", rollappPacket.Packet.GetDestChannel(), "type", rollappPacket.Type)
 		// Update the packet status
-		im.keeper.UpdateRollappPacketStatus(ctx, rollappID, rollappPacket, types.RollappPacket_ACCEPTED)
-		// Call the onRecvPacket callback for each packet
-		ack := im.app.OnRecvPacket(ctx, *rollappPacket.Packet, rollappPacket.Relayer)
-		// Write the acknowledgement to the chain only if it is synchronous
-		if ack != nil {
-			_, chanCap, err := im.keeper.LookupModuleByChannel(ctx, rollappPacket.Packet.DestinationPort, rollappPacket.Packet.DestinationChannel)
+		rollappPacket := im.keeper.UpdateRollappPacketStatus(ctx, rollappID, rollappPacket, types.RollappPacket_ACCEPTED)
+		// Call the relevant callback for each packet
+		switch rollappPacket.Type {
+		case types.RollappPacket_ON_RECV:
+			logger.Debug("Calling OnRecvPacket", "rollappID", rollappID, "sequence", rollappPacket.Packet.GetSequence(), "destination channel", rollappPacket.Packet.GetDestChannel())
+			ack := im.app.OnRecvPacket(ctx, *rollappPacket.Packet, rollappPacket.Relayer)
+			// Write the acknowledgement to the chain only if it is synchronous
+			if ack != nil {
+				wrappedFunc := func(ctx sdk.Context) error {
+					_, chanCap, err := im.keeper.LookupModuleByChannel(ctx, rollappPacket.Packet.DestinationPort, rollappPacket.Packet.DestinationChannel)
+					if err != nil {
+						return err
+					}
+					err = im.keeper.WriteAcknowledgement(ctx, chanCap, rollappPacket.Packet, ack)
+					if err != nil {
+						return err
+					}
+					return nil
+				}
+				err := osmoutils.ApplyFuncIfNoError(ctx, wrappedFunc)
+				if err != nil {
+					logger.Error("Error writing acknowledgement", "rollappID", rollappID, "sequence", rollappPacket.Packet.GetSequence(), "destination channel", rollappPacket.Packet.GetDestChannel(), "error", err.Error())
+					// Update the packet with the error
+					rollappPacket.Error = err.Error()
+					im.keeper.SetRollappPacket(ctx, rollappID, rollappPacket)
```

Comment:
why u call `SetRollappPacket` instead of Update on errors?
it will create new rollapp instance with Pending Status

Reply:
it shouldn't as the error is not part of the key generated for the packet so it should override existing one

Reply:
I think the status=pending is part of the key